src/Components/Web.JS/src/Rendering/FocusOnNavigate.ts

```diff
@@ -0,0 +1,99 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+import { domFunctions } from '../DomWrapper';
+import { EnhancedNavigationStartEvent, JSEventRegistry } from '../Services/JSEventRegistry';
+
+const customElementName = 'blazor-focus-on-navigate';
+let currentFocusOnNavigateElement: FocusOnNavigateElement | null = null;
+let allowFocusOnEnhancedLoad = false;
+
+export function enableFocusOnNavigate(jsEventRegistry: JSEventRegistry) {
+  customElements.define(customElementName, FocusOnNavigateElement);
+  jsEventRegistry.addEventListener('enhancednavigationstart', onEnhancedNavigationStart);
+  jsEventRegistry.addEventListener('enhancedload', onEnhancedLoad);
+  document.addEventListener('focusin', onFocusIn);
+  document.addEventListener('focusout', onFocusOut);
+
+  // Focus the element on the initial page load.
+  if (document.readyState === 'loading') {
+    // There may be streaming updates on the initial page load, so we want to allow
+    // those to add elements matching the active selector.
+    allowFocusOnEnhancedLoad = true;
+    document.addEventListener('DOMContentLoaded', afterInitialPageLoad);
+  } else {
+    afterInitialPageLoad();
+  }
+}
+
+function afterInitialPageLoad() {
+  tryApplyFocus(/* forceMoveFocus */ false);
+}
+
+function onEnhancedNavigationStart(ev: EnhancedNavigationStartEvent) {
+  // Only focus on enhanced load if the enhanced navigation is not a form post.
+  allowFocusOnEnhancedLoad = ev.options.method !== 'post';
+}
+
+function onEnhancedLoad() {
+  if (allowFocusOnEnhancedLoad) {
+    tryApplyFocus(/* forceMoveFocus */ true);
+  }
+}
+
+function onFocusIn() {
+  // As soon as an element get successfully focused, don't attempt to focus again on future
+  // enhanced page updates.
+  allowFocusOnEnhancedLoad = false;
+}
+
+let lastFocusedElement: Element | null = null;
+
+function onFocusOut(ev: FocusEvent) {
+  // It's possible that the element lost focus because it was removed from the page,
+  // and now the document doesn't have an active element.
+  // This could have happened either because an enhanced page update removed the element or
+  // because the focused element was replaced during the transition to interactivity
+  // (see https://github.com/dotnet/aspnetcore/issues/42561).
+  // In either case, we'll attempt to reapply focus only if:
+  // [1] that element was the one we last focused programmatically, and
+  // [2] it's about to get removed from the DOM.
+  const target = ev.target;
+  if (target instanceof Element && target === lastFocusedElement) {
+    // We want to apply focus after all synchronous changes to the DOM have completed,
+    // including the potential removal of this element.
+    setTimeout(() => {
+      if (!document.contains(target)) {
+        tryApplyFocus(/* forceMoveFocus */ false);
+      }
+    }, 0);
```

Comment:
I've done some more thorough testing and found that this isn't totally sufficient:
1. It works on Chrome and Edge, but not Firefox (the element doesn't get refocused)
2. Streaming updates may remove the `tabindex` attribute from an element that isn't otherwise focusable (like a header), and that removal unfocuses the element. We could re-apply the attribute and focus it again here, but that has additional problems:
    * This causes screen readers to re-announce the element that got focused, which can be jarring
    * That also doesn't work on Firefox 🙂 

Ideally, we would prevent the element from losing focus in the first place, when possible. I'm considering an alternative approach where we don't attempt to move focus until streaming has completed and all components have become interactive (and produced their first render). This allows us to avoid attempting to reason about what should happen if the focused element gets removed from the page. It also aligns with the behavior of an interactive `<FocusOnNavigate>` component a bit more.

Reply:
Could you clarify what scenario this addresses? Is it for the edge case where, in a sequence of streaming updates, one of them adds the desired element (and we focus it), but then a subsequent streaming update removes that element and replaces it with a different one matching the selector?

If that's the only case where this matters, is it possible we could just not do anything? As in, we focus the first one that matches during a given navigation and that's it. Developers typically want to use this feature to announce page titles (e.g., `<h1>` elems), and they won't typically be removing and re-adding page titles except in the edge case where the title text changes, and in that case there are other better ways to do it for example with `aria-live` which will announce the header text update in a way that makes sense.

Perhaps there is some other critical reason we have to handle this, but it's not obvious from the code comment. Apologies if I'm missing something specified elsewhere.

BTW great job on the testing/verification here. From your remarks I see you must have tried this with a range of browsers and at least some screen reader.

Reply:
Oh wait I think I now understand the problem:

1. First streaming update adds a `<h1>` (say), the `FocusOnNavigate` code has to add a `tabindex` to make it focusable, and then reader begins to announce it. 
2. Second streaming update removes the `tabindex` because it's not present in the original output, and the reader's announcement is truncated.

Is that right?

If so I think your idea to simply wait until enhanced nav is finished sound like the best option. That also deals with the case where the header text changes during the streaming page load. Example:

```razor
@attribute [StreamRendering]
@if (product is null)
{
    <h1>Loading...</h1>
}
else
{
    <h1>@product.Name</h1>
}
```

In this case we would want to announce `@product.Name` and not `Loading...`.

Reply:
I agree, it'll be best if we wait for the entire enhanced nav to finish before we try to focus the element. The contract in that scenario is simple, your element must be on the page after the navigation has completed. And it saves us from having to deal with all the complexity of elements being added or removed during the streaming rendering process.

Reply:
I updated the implementation to wait until an enhanced navigation completes before applying focus. However, I decided not to handle the case where a focused element gets removed upon the transition to interactivity. I started to work on this, but realized it adds quite a bit of complexity (we need to keep track of all SSR'd components currently on the page that have not transitioned to interactivity, then the interactive renderer needs to indicate somehow when each component completes its first render...).

The ultimate fix would be to not replace the DOM in the transition to interactivity (and we do have an issue tracking that). Then we wouldn't have to worry about this problem at all.